### PR TITLE
WIP: Switch to generating full RTL version

### PR DESCRIPTION
**Pull request by MohammadYounes**
This is an attempt to update the build configuration to generate a complete RTL version. Kindly check and let me know if this is okay ?

### CODE REVIEW

**Comment by simonguo** [comment]
Deploy preview for *rsuite* processing.

Building with commit be16ae78afcc641fd2f7550ad389a9733b696709

https://app.netlify.com/sites/rsuite/deploys/5dd4f274aa175700086566eb

**Comment by simonguo** [comment]
If you can build the RTL version of css into a separate file, effectively reducing the size of the default css file, I think it's fine.

@hiyangguo  Do you have any different opinions? Is there anything unreasonable?

**Comment by MohammadYounes** [comment]
@simonguo Yes, this generates both `theme-default.rtl.css` and `theme-dark.rtl.css` with the exact same size as original files.

Still, some work is required to update the direction switch button to load the appropriate file (I believe it will follow the same logic as theme switcher).

**Comment by hiyangguo** [comment]
@simonguo I think it is a good way to reduce the css file size.

**Comment by hiyangguo** [comment]
@MohammadYounes This idea is so good . But maybe you should publish a  webpack-css-rtlplugin to replace `RTLCSSPlugin.js` I think.

**Comment by MohammadYounes** [comment]
@hiyangguo A plugin already exists https://github.com/nicolashemonic/rtl-css-transform-webpack-plugin, but it doesn't work out of the box with your build setup. That's why I opted for adding a custom one.

**Comment by simonguo** [comment]
@MohammadYounes 
Just follow the way you do now.  If you need me and @hiyangguo  support, let us know.

**Comment by MohammadYounes** [comment]
@simonguo @hiyangguo check this refactoring for toggling both direction/theme 

https://github.com/MohammadYounes/rsuite.github.io/compare/wip-gen-full-rtl...MohammadYounes:wip-theme-toggle 

Any notes ?

**Comment by simonguo** [comment]
@MohammadYounes 

You also need to modify the [ready.js](https://github.com/rsuite/rsuite.github.io/blob/master/src/ready.js#L38) file to modify the css file loaded by default.

**Comment by MohammadYounes** [comment]
@simonguo Done!

**Comment by simonguo** [comment]
image
Missing the reverse of the "open chat" button

**Comment by MohammadYounes** [comment]
@simonguo Fixed.

**Comment by simonguo** [comment]
@MohammadYounes 
Awesome, this PR will be merged after rsutie@4.1.2 is released.

**Comment by MohammadYounes** [comment]
Thanks @simonguo 

One more thing, the docs (https://github.com/rsuite/rsuite.github.io/tree/master/src/pages/guide/rtl) needs to be updated to reflect the changes in this PR.

**Comment by simonguo** [comment]
@MohammadYounes  What advice do you have for the update of the rtl guide document?

**Comment by MohammadYounes** [comment]
@simonguo I meant the part related to flipping styles. 


Nevertheless, I believe it should mention the following:

1. Including the RTL version when using the compiled React Suite CSS
 `~rsuite/dist/styles/rsuite-default.rtl.css` or `~rsuite/dist/styles/rsuite-dark.rtl.css`

2. Processing the final CSS via [rtlcss](https://rtlcss.com) in case of using the Less version `~rsuite/lib/styles/themes/default/index.less` or `~rsuite/lib/styles/themes/dark/index.less`

**Comment by simonguo** [comment]
@MohammadYounes  
If you want to use the React Suite CSS, do you still need to add css build in the rsuite ?
https://github.com/rsuite/rsuite/blob/master/gulpfile.js#L36

**Comment by MohammadYounes** [comment]
@simonguo 
Yes, see https://github.com/jjlharrison/gulp-rtlcss#example

**Comment by simonguo** [comment]
@MohammadYounes 
I added gulp-rtlcss and it did not generate the correct CSS.
I think where should I use it incorrectly?
https://github.com/rsuite/rsuite/blob/chore/glup-build-rtl/gulpfile.js#L49

**Comment by MohammadYounes** [comment]
@simonguo 
All is good, just update the suffix `.pipe(rename({ suffix: '.rtl' }))`

**Comment by simonguo** [comment]
@MohammadYounes 

There is a problem with rtlcss. If there are spaces in the comment, the conversion has no effect.
image

**Comment by MohammadYounes** [comment]
@simonguo Yes, rtlcss does not ignore white spaces for value directives. 

Do you want me to update rtlcss to ignore them ? or update all used directives used in rsuite to have no preceding white space ?

**Comment by simonguo** [comment]
@MohammadYounes 
I think updating `rtlcss` might be better. This can make rtlcss have better compatibility

**Comment by MohammadYounes** [comment]
@simonguo rtlcss 2.4.1 published.

File: src/AppWithRouter.js [hunk 1]
```diff
@@ -21,37 +21,66 @@ import { createRouters } from './pages/routers';
...
+      const [theme, direction] = readTheme();
       this.state = {
-        direction
+        direction,
+        theme
       };
-      html.dir = direction;
+      //skip loading default ltr theme
+      if (theme != 'default' || direction != 'ltr')
+        this.loadTheme(theme, direction);
```
Comment (hiyangguo) [inline src/AppWithRouter.js]: There is a ESLint error , you maybe should change to 
```
 if (theme !== 'default' || direction !== 'ltr'){
        this.loadTheme(theme, direction);
}
```

File: src/AppWithRouter.js [hunk 2]
```diff
@@ -21,37 +21,66 @@ import { createRouters } from './pages/routers';
...
+      writeTheme(themeName,direction);
+      const { globalVars = {} } = window.less || {};
+      window.less &&
+        window.less.modifyVars({
+          ...globalVars,
+          '@theme-is-default': themeName === 'default'
+        });
+      html.dir = direction;
+      const themeId = getThemeId(themeName, direction);
+      loadCssFile(getThemeCssPath(themeName, direction),themeId).then(() => {
```
Comment (hiyangguo) [inline src/AppWithRouter.js]: When change style file , the page will shaking like this.

QQ20191129-103226-HD

How about added a loader in page when loading css do you think?

**Comment by hiyangguo** [comment]
@MohammadYounes When merged your PR , the grammar is failure like this:

https://github.com/rsuite/rsuite/blob/c0d1fc7392c60bbcaf363a1d52153911a8ec412f/src/Dropdown/styles/common.less#L165

In `rtl.css`
Expected value : '\EA0A';
Current value: '\EA0C';

I don't know why , Can you help me to resolve it?
You can run the code in branch [dev-4.1.2](https://github.com/rsuite/rsuite.github.io/tree/dev-4.1.2) .

**Comment by MohammadYounes** [comment]
@hiyangguo update rtlcss to 2.4.1 which ignore white space before directive prefix (see https://github.com/rsuite/rsuite.github.io/pull/161#issuecomment-559323391).

**Comment by hiyangguo** [comment]
@MohammadYounes Maybe are other reason . I'm try to update rtlcss to 2.4.1 , it is already failing.

**Comment by MohammadYounes** [comment]
@hiyangguo I'll pull the branch and check...

**Comment by MohammadYounes** [comment]
@hiyangguo This is not rtlcss related, as the value directive is lost (theme-default.css does not contain the comment).

**Comment by hiyangguo** [comment]
@MohammadYounes OK , Thank you.

**Comment by MohammadYounes** [comment]
@hiyangguo the only way I was able to keep the comment was by emitting the icon without quotes:
`content: e(@drop-right-caret-content) ~'/* rtl: "@{drop-left-caret-content}" */';`

**Comment by hiyangguo** [comment]
@MohammadYounes I don't know why will happen this strange behavior.  Maybe affected with post-css plugin I guess . But when I used [postcss-rtl](https://www.npmjs.com/package/postcss-rtl) everything is ok .And I will location this problem in a minute.

**Comment by MohammadYounes** [comment]
I believe the same problem applies, postcss-rtl is a postcss plug-in that uses rtlcss.

**Comment by hiyangguo** [comment]
@MohammadYounes But if you checkout to  the previous version everything is ok. You can try it in you local .
```
// It will drop the latest 2 commit .
git checkout 12999d49 -b old-version
```
So  I guess the problem is happened in your PR.

**Comment by MohammadYounes** [comment]
@hiyangguo The difference is in the processing pipeline, when you were using postcss-rtl, you operated inside the same postcss pipeline, not on the final file.  

My guess the comment gets stripped in `css-loader`, after postcss processing is done.

**Comment by MohammadYounes** [comment]
@hiyangguo I managed to pinpoint the root cause of the issue, as I guessed, the comment is lost in `css-loader`. 

# The Fix
Make all [icon codes](https://github.com/rsuite/rsuite/blob/master/src/styles/iconfont-variables.less) upper-case `\EA0A` instead of `\ea0a`.

## Explanation
The following code exists in `css-loader`

```js
css.walkDecls(function(decl) {
  var values = Tokenizer.parseValues(decl.value);
  values.nodes.forEach(function(value) {
    value.nodes.forEach(processNode);
  });
  decl.value = Tokenizer.stringifyValues(values);
});
```
It parses the value then updates the declaration value using the stringified version. This leads to postcss truncating the parsed raw data [`decl.raws`](http://api.postcss.org/Node.html#raws) containing the comment text.  Why? because the value was changed, as the original value is `\ea0a` and the stringified value is `\EA0A`.

You can test the above using the following sample:

```js
var postcss = require('postcss');
var Tokenizer = require('css-selector-tokenizer');

var parserPlugin = postcss.plugin('css-loader-parser', function(options) {
  return function(css) {
    console.log(
      `-------------pre------------- ${css} --------------------------`
    );

    css.walkDecls(function(decl) {
      var values = Tokenizer.parseValues(decl.value);
      console.log(`
      decl.value: ${decl.value}
      stringified value: ${Tokenizer.stringifyValues(values)};
      `);
      decl.value = Tokenizer.stringifyValues(values);
    });

    console.log(
      `-------------post------------- ${css} --------------------------`
    );
  };
});

postcss([parserPlugin()]).process(`
 .lower-case:before {
    content: '\\f060' /*!rtl:'\\f061' */;
  }
  .upper-case:before {
    content: '\\F060' /*!rtl:'\\f061' */;
  }
`).css;
```

Thanks.

**Comment by hiyangguo** [comment]
@MohammadYounes Thank you very much , I spent more than 5 hours to find reason of this issue.It is really affect by `css-loader`. And I updated `css-loader@latest` . [diff](https://github.com/rsuite/rsuite.github.io/pull/167/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R104)
```diff
 -   "css-loader": "^0.28.11",
 +   "css-loader": "^3.2.1",
```
Then everything is OK. So #764 is no use , right ?

**Comment by MohammadYounes** [comment]
@hiyangguo I didn't pay attention to `css-loader` version. I agree, no need for that PR anymore.

**Comment by hiyangguo** [comment]
@MohammadYounes Can you help me to resolve the sourcemap ? I can't generated it , but when  I developing , I need it very much .You can help me in [rsuite docs RTLCSSPlugin.js](https://github.com/rsuite/rsuite/blob/master/docs/scripts/RTLCSSPlugin.js)。 Thank you very much.

**Comment by MohammadYounes** [comment]
@hiyangguo [`rtlcss.process`](https://github.com/MohammadYounes/rtlcss/blob/master/lib/rtlcss.js#L177) is a shortcut to get the processed CSS only. To get source maps, you need to [use it as postcss plugin](https://rtlcss.com/learn/usage-guide/install/#advanced-usage):

```js
postcss(require('rtlcss')).process(content).then(result => {
  result.css // Processed CSS
  result.map // Source map
})
```

**Pull request merged**